### PR TITLE
Add EIP-5299: Storage scaling pattern

### PR DESCRIPTION
**Pull request by v9hstk**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-5299.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+eip: 5299
+title: Layer3 scaling pattern
+description: A primitive to facilitate eternal upgradeability of smart contracts, by making storage slots dynamic.
+author: Vinay Kumar (@v9hstk), Pratik Agarwal (@Prtk418)
+discussions-to: https://ethereum-magicians.org/t/eip-9000-layer3-scaling-standard/9039
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-07-18
+---
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:11]:
```suggestion
requires: 2535
---
```

Requires is a misnomer; it should be named references.

File: EIPS/eip-5299.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+## Abstract
+Upgrading a storage layout is required in case of a complex project while it’s re architected. The EIP introduces implementation of dynamic storage slots, which can be upgraded to a new one in case of an upgrade. Without actually deploying a new contract altogether.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:14]:
```suggestion
Upgrading a storage layout is required in case of a complex project if it is re-architected. The EIP introduces dynamic storage slots, which can be upgraded without deploying a new contract.
```

File: EIPS/eip-5299.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+## Motivation
+With the existing smart contract upgradable design patterns available, while it’s possible to upgrade the code implementation, it is not possible to completely upgrade storage structure. Which leads to re deploying the contracts and introduces a new version of the app
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:17]:
```suggestion
With the existing smart contract upgradable design patterns available, while it’s possible to upgrade the code implementation, it is not possible to completely upgrade the storage structure, which requires a re-deployment of the contracts.
```

File: EIPS/eip-5299.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+From end user perspective, it is a friction to migrate from old version of app to a new version because of multiple transactions involved
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:19]:
```suggestion
From the end-user perspective, it is annoying to migrate from an old version of an app to a new version, especially if it requires deploying multiple transactions.
```

File: EIPS/eip-5299.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+What if we could use the old storage in conjunction with the new storage structure, or we could migrate a user data to newer structure very easily in a single transaction.
+
+With the implementation of this EIP, it would become very convenient to upgrade storage which is a huge problem currently.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:23]:
```suggestion
```

File: EIPS/eip-5299.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+## Specification
+The specification is an addition to the standard ([EIP-2535](./eip-2535.md)) diamond standard design which focuses on putting forward a way for implementing upgradable contracts where the implementation logic is modular where the parts of a contract can be added/replaced/removed while leaving other parts alone. But when it comes to upgrading the storage structure we need a mechanism so that we can maintain the old data points as well and update the implementation in a way that its backward compatible by using the deprecated storage and either migrate to newer storage or utilise both the deprecated storage and the upgraded one.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:26]:
```suggestion
This EIP is an extension of [EIP-2535](./eip-2535.md), adding a mechanism that maintains the old data points while updating the implementation.
```

File: EIPS/eip-5299.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+Before diving into specifics about the storage upgradability, let’s understand how diamond standard works.
+
+### How diamond standard works
+The exposed contract stores within it a mapping of function selector to implementation contract address. When an external function is called on the exposed contract its fallback function is executed. The fallback function finds which implementation contract has the function that has been called and then executes that function from the implementation contract using delegatecall. The exposed contract’s fallback function and delegatecall enable the contract to execute implementation contract’s external function as its own external function. The msg.sender and msg.value values do not change and only the exposed contract’s storage is read and written to.
+
+The exposed contract can use a function to add/replace/remove any number of functions from any number of implementation contract in a single transaction. This function updates the mapping of function selector to implementation contract address.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:33]:
```suggestion
```
Not needed. Readers should have already read EIP-2535

**Review by Pandapip1 — COMMENTED** [review]
Here's my review so far. This is incomplete.

**Comment by v9hstk** [comment]
> Here's my review so far. This review is incomplete, as I need to leave to go somewhere.

Thanks @Pandapip1 
We've amended the suggestions.

File: EIPS/eip-5299.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+## Backwards Compatibility
+This standard makes scaling smart contract storage compatible with the future standards & functionality because the storage slots can be added dynamically to enable storage expansion, while also facilitating a means to retrieve and store data across various legacy & in-production storage contracts.
+
+
+## Test Cases
+Test cases are included in the reference implementation refer to Layer3-scaling-standard repository in 0xHashstack's github account
+
+
+## Reference Implementation
+Refer to Layer3-scaling-standard repository in 0xHashstack's github account
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:157]: Must be put in the assets folder.

File: EIPS/eip-5299.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+
+## Security Considerations
+A storage selector clash occurs when two different storage slot address points to the same location. This has the unintended consequence of overriding data which is unintentional. This scenario is not possible with the diamondCut function prescribed in ([EIP-2535](./eip-2535.md)), because it uses the contract address of facets to generate the slot address.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:161]:
```suggestion
A storage selector clash occurs when two different storage slot addresses point to the same location. This has the unintended consequence of overriding data. This scenario is not possible with the `diamondCut` function described in ([EIP-2535](./eip-2535.md)), because it uses the contract address of facets to generate the slot address.
```

File: EIPS/eip-5299.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+## Rationale
+In order to make the upgrade process as smooth and seamless possible it is mandatory to keep all the components involved in a system as modular as possible. Hence the upgradable storage structure will prove to be very helpful.
+
+
+## Backwards Compatibility
+This standard makes scaling smart contract storage compatible with the future standards & functionality because the storage slots can be added dynamically to enable storage expansion, while also facilitating a means to retrieve and store data across various legacy & in-production storage contracts.
+
+
+## Test Cases
+Test cases are included in the reference implementation refer to Layer3-scaling-standard repository in 0xHashstack's github account
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:153]: Must be put in the assets folder.

File: EIPS/eip-5299.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        LibB.AppStorageV2 storage as2 = LibA.appStorageV2();
+        return (as1.dataA, as2.dataA);
+    }
+}
+
+```
+With above implementation, we will have access to both AppStorageV1 and AppStorageV2 which can be used either for migration or used in conjunction as per the use case.
+
+## Rationale
+In order to make the upgrade process as smooth and seamless possible it is mandatory to keep all the components involved in a system as modular as possible. Hence the upgradable storage structure will prove to be very helpful.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:145]:
```suggestion
WIP
```

A rationale explains design decisions made in the Specification (e.g. why a function is named like it is, why a specific type is used instead of another, etc...)

While it is a required section, placeholders for this specific section are common enough that I would no longer consider having a full Rationale necessary for an EIP to be merged as Draft.

File: EIPS/eip-5299.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+
+## Specification
+This EIP is an extension of [EIP-2535](./eip-2535.md), adding a mechanism that maintains the old data points while updating the implementation.
+
+
+### Layer3 scaling standard
+For upgrading a storage structure we will store the slot addresses in a special struct we will call it ParentStorage and then retrieve the slot whenever the specific facet interacts with data store.
+
+Here is a simple example that shows dynamic slot based storage and its use in one of the implementation contract:
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:30]:
```suggestion
Here is a simple example that shows dynamic slot based storage and its use in one of the implementation contract:

```

File: EIPS/eip-5299.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+```
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:31]:
```suggestion
```solidity
```

File: EIPS/eip-5299.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        as.dataA = _dataA;
+    }
+
+    function getDataA() external view returns (bytes32) {
+        LibA.AppStorageV1 storage as = LibA.appStorageV1();
+        return as.dataA;
+    }
+}
+```
+Now that we have unique slot position for our implementation contract in place, we will modify the storage structure and introduce a new slot position for the same and upgrade the contract by using a function which will add the new function from below implementation contract. The function updates the mapping of function selector to implementation contract address.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:77]:
```suggestion

Now that we have unique slot position for our implementation contract in place, we will modify the storage structure and introduce a new slot position for the same and upgrade the contract by using a function which will add the new function from below implementation contract. The function updates the mapping of function selector to implementation contract address.

```

File: EIPS/eip-5299.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+```
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:78]:
```suggestion
```solidity
```

File: EIPS/eip-5299.md [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+    function getDataAV1AndV2() external view returns (bytes32, uint256) {
+        LibB.AppStorageV1 storage as1 = LibA.appStorageV1();
+        LibB.AppStorageV2 storage as2 = LibA.appStorageV2();
+        return (as1.dataA, as2.dataA);
+    }
+}
+
+```
+With above implementation, we will have access to both AppStorageV1 and AppStorageV2 which can be used either for migration or used in conjunction as per the use case.
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:142]:
```suggestion
```
Implied.

**Review by MicahZoltu — COMMENTED** [review]
This EIP feels like a Good Idea, rather than a standard.  EIPs generally should be things that need to be standardized, and it isn't really an appropriate place to author/store design patterns or other Good Ideas.

For example, a standard should essentially never have private/internal methods, it should only describe the public surface area and the invariants that can be expected to be maintained by the caller of those methods.  This proposal is almost entirely made up of internal/private members suggesting that it is a Good Idea/Design Pattern, not a standard.

Another way to look at this is to ask whether there is a many-to-many relationship between callers and callees or not.  For Twitter API, there is one server serving the data and many users which means a standard is not necessary as Twitter can just publish API docs.  For web browsers on the other hand, there are many web browsers and many web pages and we want all browsers to be able to render all pages, so we need standards.

File: EIPS/eip-5218.md [hunk 17]
```diff
@@ -0,0 +1,237 @@
+---
+eip: 5218
```
Comment (Pandapip1) [inline EIPS/eip-5218.md:2]: ?!!??!?!?! Why is this here?

Comment (prtk418) [reply EIPS/eip-5218.md]: This was because branch got outdated, merged with latest 👍

File: assets/eip-5299/contracts/LICENSE [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
```
Comment (MicahZoltu) [inline assets/eip-5299/contracts/LICENSE:21]: Everything in this repository must be CC0 licensed.

File: assets/eip-5299/contracts/README.md [hunk 19]
```diff
@@ -0,0 +1,6 @@
+# EIP-9000
+Layer3 scaling standard.
+
+The proposed EIP is published at Ethereum Magicians for discussions: https://ethereum-magicians.org/t/eip-9000-layer3-scaling-standard/9039
+
+This is an implementation for [EIP-9000 Layer3 scaling standard](https://ethereum-magicians.org/t/eip-9000-layer3-scaling-standard/9039).
```
Comment (MicahZoltu) [inline assets/eip-5299/contracts/README.md:6]: This is redundant with the content of the EIP and shouldn't be included.

File: assets/eip-5299/contracts/package.json [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+    "chai": "^4.3.4",
+    "dotenv": "^8.6.0",
+    "ethereum-waffle": "^3.4.0",
+    "ethers": "^5.4.1",
+    "hardhat": "^2.4.1",
+    "prettier": "^2.3.2",
+    "prettier-plugin-solidity": "^1.0.0-beta.13",
+    "standard": "^16.0.3"
+  }
+}
```
Comment (MicahZoltu) [inline assets/eip-5299/contracts/package.json:31]: Build tooling isn't appropriate for an EIP.  It would be appropriate for a tutorial, your own GitHub repository, a blog, etc.  Plus, external links not allowed and package.json is basically by definition a bunch of external links.

File: EIPS/eip-5299.md [hunk 21]
```diff
@@ -0,0 +1,168 @@
+---
+eip: 5299
+title: Layer3 scaling pattern
```
Comment (MicahZoltu) [inline EIPS/eip-5299.md:3]: While we normally don't curate EIP titles, I'm a bit uncomfortable with this title given how widely used the term `Layer 2` is in the Ethereum ecosystem, and calling this `Layer 3 Scaling` may imply to the reader that this is some kind of overarching design pattern we are working toward rather than a mechanism for dealing with contract storage.

Also, the name doesn't seem at all related to what this EIP actually purports to do (contract upgrade pattern).

File: EIPS/eip-5299.md [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+From the end-user perspective, it is annoying to migrate from an old version of an app to a new version, especially if it requires deploying multiple transactions.
+
+
+## Specification
+This EIP is an extension of [EIP-2535](./eip-2535.md), adding a mechanism that maintains the old data points and enabling the possibility to update the structure of storage struct while updating the implementation.
+
+### Layer3 scaling standard
+For upgrading a storage structure we will store the slot addresses in a special struct we will call it ParentStorage and then retrieve the slot whenever the specific facet interacts with data store.
+
+Here is a simple example that shows dynamic slot based storage and its use in one of the implementation contract:
```
Comment (MicahZoltu) [inline EIPS/eip-5299.md:29]: I don't see an actual specification here.  The majority of this section appears to be an example implementation (which should go in the `## Reference Implementation` section) but no actual specification exists at the moment and that is required for an EIP.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
This doesn't feel like it needs to be an EIP.  Design patterns fall into the category of Good Ideas, which is quite different from Standards.  Standards are useful when you have many to many relationships between components of a system that need to communicate, while Good Ideas can be implemented by a project without needing to worry about compatibility with other projects.

A good indicator that something is a Good Idea rather than a Standard is when everything is private, internal, or permissioned (as is the case in this EIP save for one getter).  If you *do* want to standardize the public surface area, then the specification should be trimmed down to *only* the publicly accessible surface area that all implementations must conform to.

-----

The assets associated with an EIP should not be a full application in general, it should just be enough code so that someone reading the EIP can get an idea of what an implementation might look like.  Currently, this EIP includes a *huge* amount of code when it should maybe include a single contract that implements an interface.

The same goes for test cases.  A good set of test cases define input parameters and expected output or other publicly visible results, but they shouldn't (generally) be in the form of a complete executable set of tests.  There are sometimes some exceptions to this, but I don't think they apply here.

Deploy scripts *definitely* shouldn't be part of the EIP, nor should things like hardhat/JS configuration files.

File: EIPS/eip-5299.md [hunk 23]
```diff
@@ -0,0 +1,168 @@
+---
+eip: 5299
+title: Layer3 scaling pattern
+description: A primitive to facilitate eternal upgradeability of smart contracts, by making storage slots dynamic.
+author: Vinay Kumar (@v9hstk), Pratik Agarwal (@Prtk418)
+discussions-to: https://ethereum-magicians.org/t/eip-9000-layer3-scaling-standard/9039
+status: Draft
+type: Standards Track
+category: ERC
```
Comment (Pandapip1) [inline EIPS/eip-5299.md:9]:
```suggestion
type: Informational
```

Comment (MicahZoltu) [reply EIPS/eip-5299.md:9]: I'm not a fan of turning `Informational` EIPs into a "collection of Best Practices and Good Ideas".  At best, we'll end up with a mix of some good ideas, and some bad ideas since we don't have any means of curation (nor do I think editors should get into the game of Good Idea curation).

Comment (v9hstk) [reply EIPS/eip-5299.md:9]: > ```suggestion
> 
> type: Informational
> 
> ```

Please refer to the implementation examples provided in the assets. 

The key value add this design pattern enables is to induce composability in the storage structure, a key missing component in all the upgradable patterns. 

I will add additional implementation references describing storage scalability in low level upgradable patterns.

File: EIPS/eip-5299.md (hunk 21 above)
Comment (v9hstk) [reply EIPS/eip-5299.md:3]: Thanks for your comments on our Ethereum Improvement Proposal. Everyone is entitled to their opinion irrespective of its relevance, or wisdom.

If you have specific feedback on the proposed design short-comings, such as probable oversights, or edge cases that the developers might have overseen, me & my team of developers are happy to look into it.

If you have trouble understanding the scale & scope of value such a simplistic looking design pattern could bring to the ecosystem, you can look at Uniswap v1,v2,v3 as an example of the world that's deprived of this EIP. If you still have trouble, i'd happy to connect with you over a call, or probably a twitter space even to dissect the proposal in greater detail.

File: EIPS/eip-5299.md (hunk 23 above)
Comment (v9hstk) [reply EIPS/eip-5299.md:9]: @MicahZoltu Thanks for your comments on our Ethereum Improvement Proposal. Everyone is entitled to their opinion irrespective of its relevance, or wisdom.

If you have specific feedback on the proposed design short-comings, such as probable oversights, or edge cases that the developers might have overseen, me & my team of developers are happy to look into it.

If you have trouble understanding the scale & scope of value-add such a simplistic looking design pattern could bring to the ecosystem, you can look at Uniswap v1,v2,v3 as an example of the world that's deprived of this EIP. All upgradable patterns today enable functional and/or logic upgradability. By solving for storage scalability, we are able to make upgradable patterns lot more composable and enable a means to truly scale smart contracts.

 If you still have trouble, I'm happy to connect with you over a call, or probably a twitter space even, to dissect the proposed EIP in greater detail.

File: EIPS/eip-5299.md (hunk 21 above)
Comment (MicahZoltu) [reply EIPS/eip-5299.md:3]: As editors, it is our job to ensure that content in new EIPs is appropriate content for an EIP.  We also work to ensure that EIPs follow fairly consistent standards and practices as much as possible.  While I can appreciate that many people think that the scope of EIPs should be broader than what it currently is (standards), at the moment this repository is primarily for standards, not design patterns and best practices.

**Comment by v9hstk** [comment]
> This doesn't feel like it needs to be an EIP. Design patterns fall into the category of Good Ideas, which is quite different from Standards. Standards are useful when you have many to many relationships between components of a system that need to communicate, while Good Ideas can be implemented by a project without needing to worry about compatibility with other projects.
> 
> A good indicator that something is a Good Idea rather than a Standard is when everything is private, internal, or permissioned (as is the case in this EIP save for one getter). If you _do_ want to standardize the public surface area, then the specification should be trimmed down to _only_ the publicly accessible surface area that all implementations must conform to.
> 
> The assets associated with an EIP should not be a full application in general, it should just be enough code so that someone reading the EIP can get an idea of what an implementation might look like. Currently, this EIP includes a _huge_ amount of code when it should maybe include a single contract that implements an interface.
> 
> The same goes for test cases. A good set of test cases define input parameters and expected output or other publicly visible results, but they shouldn't (generally) be in the form of a complete executable set of tests. There are sometimes some exceptions to this, but I don't think they apply here.
> 
> Deploy scripts _definitely_ shouldn't be part of the EIP, nor should things like hardhat/JS configuration files.

This is a basic restructuring suggestion.  Will be updated tomorrow.

File: EIPS/eip-5299.md (hunk 21 above)
Comment (Pandapip1) [reply EIPS/eip-5299.md:3]: I agree with micah here.

File: assets/eip-5299/contracts/package.json (hunk 20 above)
Comment (Pandapip1) [inline assets/eip-5299/contracts/package.json:31]:
```suggestion
```

File: assets/eip-5299/contracts/LICENSE (hunk 18 above)
Comment (Pandapip1) [inline assets/eip-5299/contracts/LICENSE:21]:
```suggestion
```

File: assets/eip-5299/contracts/README.md (hunk 19 above)
Comment (Pandapip1) [inline assets/eip-5299/contracts/README.md:6]:
```suggestion
```

File: assets/eip-5299/contracts/hardhat.config.js [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+ */
+module.exports = {
+  solidity: '0.8.6',
+  settings: {
+    optimizer: {
+      enabled: true,
+      runs: 200
+    }
+  }
+}
```
Comment (Pandapip1) [inline assets/eip-5299/contracts/hardhat.config.js:29]:
```suggestion
```

File: assets/eip-5299/contracts/scripts/deploy.js [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+if (require.main === module) {
+  deployDiamond()
+    .then(() => process.exit(0))
+    .catch(error => {
+      console.error(error)
+      process.exit(1)
+    })
+}
+
+exports.deployDiamond = deployDiamond
```
Comment (Pandapip1) [inline assets/eip-5299/contracts/scripts/deploy.js:81]:
```suggestion
```

File: assets/eip-5299/contracts/scripts/libraries/diamond.js [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+  }
+}
+
+exports.getSelectors = getSelectors
+exports.getSelector = getSelector
+exports.FacetCutAction = FacetCutAction
+exports.StorageAction = StorageAction
+exports.remove = remove
+exports.removeSelectors = removeSelectors
+exports.findAddressPositionInFacets = findAddressPositionInFacets
```
Comment (Pandapip1) [inline assets/eip-5299/contracts/scripts/libraries/diamond.js:84]:
```suggestion
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Remove the tooling files, please.

File: EIPS/eip-5299.md [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+## Abstract
+Upgrading a storage layout is required in case of a complex project if it is re-architected. The EIP introduces dynamic storage slots, which can be upgraded without deploying a new contract.
+
+## Motivation
+With the existing smart contract upgradable design patterns available, while it’s possible to upgrade the code implementation, it is not possible to completely upgrade the storage structure, which requires a re-deployment of the contracts.
+
+From the end-user perspective, it is annoying to migrate from an old version of an app to a new version, especially if it requires deploying multiple transactions.
+
+
+## Specification
```
Comment (xinbenlv) [inline EIPS/eip-5299.md:23]: Thank you for the EIP! Please consider complete the spec section, otherwise it's hard for readers to understand what kind of practice / interface this EIP proposes other people to follow.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: assets/eip-5299/contracts/upgradeInitializers/DiamondInit.sol [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+
+import {LibDiamond} from "../libraries/LibDiamond.sol";
+import { IDiamondLoupe } from "../interfaces/IDiamondLoupe.sol";
+import { IDiamondCut } from "../interfaces/IDiamondCut.sol";
+import { IERC173 } from "../interfaces/IERC173.sol";
+import { IERC165 } from "../interfaces/IERC165.sol";
+
+// It is expected that this contract is customized if you want to deploy your diamond
+// with data from a deployment script. Use the init function to initialize state variables
+// of your diamond. Add parameters to the init funciton if you need to.
```
Comment (SamWilsn) [inline assets/eip-5299/contracts/upgradeInitializers/DiamondInit.sol:19]: Typo here

**Comment by genzmiracle6800** [comment]
Innovation 🚀
